diff --git a/core/misc/ajax.js b/core/misc/ajax.js
index 5ea5242..64dde52 100644
--- a/core/misc/ajax.js
+++ b/core/misc/ajax.js
@@ -483,10 +483,6 @@ function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr
       var $new_content_wrapped = $('<div></div>').html(response.data);
       var $new_content = $new_content_wrapped.contents();
 
-      if ($new_content.length !== 1 || $new_content.get(0).nodeType !== 1) {
-        $new_content = $new_content_wrapped;
-      }
-
       switch (method) {
         case 'html':
         case 'replaceWith':
@@ -513,6 +509,9 @@ function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr
 
       if ($new_content.parents('html').length > 0) {
         settings = response.settings || ajax.settings || drupalSettings;
+        if ($new_content.length !== 1 || $new_content.get(0).nodeType !== 1) {
+          $new_content = $new_content.parent();
+        }
         Drupal.attachBehaviors($new_content.get(0), settings);
       }
     },
