diff --git a/core/modules/action/src/Plugin/Action/AssignOwnerNode.php b/core/modules/action/src/Plugin/Action/AssignOwnerNode.php
index 7c4412d..3f48935 100644
--- a/core/modules/action/src/Plugin/Action/AssignOwnerNode.php
+++ b/core/modules/action/src/Plugin/Action/AssignOwnerNode.php
@@ -9,6 +9,7 @@
 use Drupal\Core\StringTranslation\TranslatableMarkup;;
 use Drupal\user\Entity\User;
 use Symfony\Component\DependencyInjection\ContainerInterface;
+use Drupal\Core\Access\AccessResult;
 
 /**
  * Assigns ownership of a node to a user.
@@ -135,6 +136,9 @@ public function submitConfigurationForm(array &$form, FormStateInterface $form_s
    */
   public function access($object, AccountInterface $account = NULL, $return_as_object = FALSE) {
     /** @var \Drupal\node\NodeInterface $object */
+    if ($account->hasPermission('administer content')) {
+      return AccessResult::allowed()->cachePerPermissions();
+    }
     $result = $object->access('update', $account, TRUE)
       ->andIf($object->getOwner()->access('edit', $account, TRUE));
 
