diff --git a/core/lib/Drupal/Core/Utility/LinkGenerator.php b/core/lib/Drupal/Core/Utility/LinkGenerator.php
index 8367556445..e10fcd8197 100644
--- a/core/lib/Drupal/Core/Utility/LinkGenerator.php
+++ b/core/lib/Drupal/Core/Utility/LinkGenerator.php
@@ -150,7 +150,7 @@ public function generate($text, Url $url) {
 
     // Move attributes out of options since generateFromRoute() doesn't need
     // them. Make sure the "href" comes first for testing purposes.
-    $attributes = ['href' => ''] + $variables['options']['attributes'];
+    $attributes = array_merge(['href' => ''], $variables['options']['attributes']);
     unset($variables['options']['attributes']);
     $url->setOptions($variables['options']);
 
